File: Php/entities/CashgameResultFactoryImpl.php
<?php
namespace entities{

	use DateTime;
	use entities\Checkpoints\Checkpoint;

	class CashgameResultFactoryImpl implements CashgameResultFactory{

		public function __construct(){

		}

		public function create(Player $player, array $checkpoints){
			$result = new CashgameResult();

			$buyin = $this->getBuyinSum($checkpoints);
			$stack = $this->getStack($checkpoints);
			$winnings = $stack - $buyin;
			$buyinTime = $this->getBuyinTime($checkpoints);
			$lastReportTime = $this->getLastReportTime($checkpoints);
			$cashoutCheckpoint = $this->getCashoutCheckpoint($checkpoints);
			$cashoutTime = $cashoutCheckpoint != null ? $cashoutCheckpoint->getTimeStamp() : null;
			$playedTime = $this->getPlayedTime($buyinTime, $cashoutTime);
			$hasReported = $this->hasReported($checkpoints);

			$result->setPlayer($player);
			$result->setCheckpoints($checkpoints);
			$result->setBuyin($buyin);
			$result->setStack($stack);
			$result->setWinnings($winnings);
			$result->setBuyinTime($buyinTime);
			$result->setCashoutTime($cashoutTime);
			$result->setLastReportTime($lastReportTime);
			$result->setPlayedTime($playedTime);
			$result->setCashoutCheckpoint($cashoutCheckpoint);
			$result->setHasReported($hasReported);

			return $result;
		}

		/**
		 * @param Checkpoint[] $checkpoints
		 * @return int
		 */
		private function getBuyinSum(array $checkpoints){
			$typedCheckpoints = $this->getCheckpointsOfType($checkpoints, 'entities\Checkpoints\BuyinCheckpoint');
			$buyin = 0;
			foreach($typedCheckpoints as $checkpoint){
				$buyin += $checkpoint->getAmount();
			}
			return $buyin;
		}

		/**
		 * @param Checkpoint[] $checkpoints
		 * @param $type
		 * @return Checkpoint[]
		 */
		private function getCheckpointsOfType(array $checkpoints, $type){
			$typedCheckpoints = array();
			foreach($checkpoints as $checkpoint){
				if(get_class($checkpoint) == $type){
					$typedCheckpoints[] = $checkpoint;
				}
			}
			return $typedCheckpoints;
		}

		private function getStack(array $checkpoints){
			$checkpoint = $this->getLastCheckpoint($checkpoints);
			return $checkpoint != null ? $checkpoint->getStack() : 0;
		}

		/**
		 * @param Checkpoint[] $checkpoints
		 * @return CheckPoint|null
		 */
		private function getLastCheckpoint(array $checkpoints){
			return count($checkpoints) != 0 ? end($checkpoints) : null;
		}

		private function getBuyinTime(array $checkpoints){
			$checkpoint = $this->getFirstBuyinCheckpoint($checkpoints);
			if($checkpoint == null){
				return null;
			} else {
				return $checkpoint->getTimestamp();
			}
		}

		private function getCashoutTime(array $checkpoints){
			$checkpoint = $this->getCashoutCheckpoint($checkpoints);
			if($checkpoint == null){
				return null;
			}
			return $checkpoint->getTimestamp();
		}

		private function getFirstBuyinCheckpoint(array $checkpoints){
			return $this->getCheckpointOfType($checkpoints, 'entities\Checkpoints\BuyinCheckpoint');
		}

		private function getCashoutCheckpoint(array $checkpoints){
			return $this->getCheckpointOfType($checkpoints, 'entities\Checkpoints\CashoutCheckpoint');
		}

		/**
		 * @param Checkpoint[] $checkpoints
		 * @param $type
		 * @return Checkpoint|null
		 */
		private function getCheckpointOfType(array $checkpoints, $type){
			foreach($checkpoints as $checkpoint){
				if(get_class($checkpoint) == $type){
					return $checkpoint;
				}
			}
			return null;
		}

		private function getPlayedTime(DateTime $startTime = null, DateTime $endTime = null){
			if($startTime == null || $endTime == null){
				return 0;
			}
			$timespan = $startTime->diff($endTime);
			return $timespan->h * 60 + $timespan->i;
		}

		private function getLastReportTime(array $checkpoints){
			$checkpoint = $this->getLastCheckpoint($checkpoints);
			if($checkpoint == null){
				return new DateTime();
			} else {
				return $checkpoint->getTimestamp();
			}
		}

		public function hasReported(array $checkpoints){
			$reportCheckpoints = $this->getCheckpointsOfType($checkpoints, 'entities\Checkpoints\ReportCheckpoint');
			return count($reportCheckpoints) > 0;
		}

	}

}

File: PokerBunch.Infrastructure/Factories/CashgameResultFactoryImpl.cs
using System;
using System.Collections.Generic;
using Core.Classes;
using Core.Classes.Checkpoints;
using entities;

namespace Infrastructure.Factories
{
    public class CashgameResultFactoryImpl : CashgameResultFactory
    {

        public CashgameResult Create(Player player, List<Checkpoint> checkpoints)
        {
            var result = new CashgameResult();

            var buyin = GetBuyinSum(checkpoints);
            var stack = GetStack(checkpoints);
            var winnings = stack - buyin;
            var buyinTime = GetBuyinTime(checkpoints);
            var lastReportTime = GetLastReportTime(checkpoints);
            var cashoutCheckpoint = GetCashoutCheckpoint(checkpoints);
            DateTime? cashoutTime = null;
            if (cashoutCheckpoint != null)
                cashoutTime = cashoutCheckpoint.Timestamp;
            var playedTime = GetPlayedTime(buyinTime, cashoutTime);
            var hasReported = HasReported(checkpoints);

            result.Player = player;
            result.Checkpoints = checkpoints;
            result.Buyin = buyin;
            result.Stack = stack;
            result.Winnings = winnings;
            result.BuyinTime = buyinTime;
            result.CashoutTime = cashoutTime;
            result.LastReportTime = lastReportTime;
            result.PlayedTime = playedTime;
            result.CashoutCheckpoint = cashoutCheckpoint;
            result.HasReported = hasReported;

            return result;
        }

        private int GetBuyinSum(List<Checkpoint> checkpoints)
        {
            var buyinCheckpoints = GetCheckpointsOfType(checkpoints, typeof (BuyinCheckpoint));
            var buyin = 0;
            foreach (var checkpoint in buyinCheckpoints)
            {
                buyin += checkpoint.Amount;
            }
            return buyin;
        }

        private List<Checkpoint> GetCheckpointsOfType(List<Checkpoint> checkpoints, Type type)
        {
            var typedCheckpoints = new List<Checkpoint>();
            foreach (var checkpoint in checkpoints)
            {
                if (checkpoint.GetType() == type)
                {
                    typedCheckpoints.Add(checkpoint);
                }
            }
            return typedCheckpoints;
        }

        private int GetStack(List<Checkpoint> checkpoints)
        {
            var checkpoint = GetLastCheckpoint(checkpoints);
            return checkpoint != null ? checkpoint.Stack : 0;
        }

        private Checkpoint GetLastCheckpoint(List<Checkpoint> checkpoints)
        {
            return checkpoints.Count > 0 ? checkpoints[checkpoints.Count - 1] : null;
        }

        private DateTime? GetBuyinTime(List<Checkpoint> checkpoints)
        {
            var checkpoint = GetFirstBuyinCheckpoint(checkpoints);
            if (checkpoint == null)
            {
                return null;
            }
            return checkpoint.Timestamp;
        }
        
        private Checkpoint GetFirstBuyinCheckpoint(List<Checkpoint> checkpoints)
        {
            return GetCheckpointOfType(checkpoints, typeof (BuyinCheckpoint));
        }

        private Checkpoint GetCashoutCheckpoint(List<Checkpoint> checkpoints)
        {
            return GetCheckpointOfType(checkpoints, typeof (CashoutCheckpoint));
        }

        private Checkpoint GetCheckpointOfType(List<Checkpoint> checkpoints, Type type)
        {
            foreach (var checkpoint in checkpoints)
            {
                if (checkpoint.GetType() == type)
                {
                    return checkpoint;
                }
            }
            return null;
        }

        private int GetPlayedTime(DateTime? startTime = null, DateTime? endTime = null)
        {
            if (!startTime.HasValue || !endTime.HasValue)
            {
                return 0;
            }
            var timespan = endTime - startTime;
            return (int)Math.Round(timespan.Value.TotalMinutes);
        }

        private DateTime GetLastReportTime(List<Checkpoint> checkpoints)
        {
            var checkpoint = GetLastCheckpoint(checkpoints);
            if (checkpoint == null)
            {
                return new DateTime();
            }
            return checkpoint.Timestamp;
        }

        public bool HasReported(List<Checkpoint> checkpoints)
        {
            var reportCheckpoints = GetCheckpointsOfType(checkpoints, typeof (ReportCheckpoint));
            return reportCheckpoints.Count > 0;
        }

    }

}